feat(extension): add Spring Boot extension (#2337)
canonical/charmcraft: docs/tutorial/kubernetes-charm-spring-boot.rst

  4 GB RAM, and a 50 GB disk.
- Familiarity with Linux.

What you'll do
--------------

#. Create a Spring Boot app.
#. Use that to create a rock with Rockcraft.
#. Use that to create a charm with Charmcraft.
#. Use that to test, deploy, configure, etc., your Spring Boot app on a local
   Kubernetes cloud with Juju.
#. Repeat the process, mimicking a real development process.

.. important::

    Should you get stuck or notice issues, please get in touch on
    `Matrix <https://matrix.to/#/#12-factor-charms:ubuntu.com>`_ or
    `Discourse <https://discourse.charmhub.io/>`_

Set things up
-------------

.. include:: /reuse/tutorial/setup_edge.rst
.. |12FactorApp| replace:: Spring Boot

As the ``spring-boot-framework`` extensions for Rockcraft and Charmcraft are
still in development, we must enable experimental extensions for each:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:export-experimental-env-vars]
    :end-before: [docs:export-experimental-env-vars-end]
    :dedent: 2

Create the Spring Boot app
--------------------------

Start by creating the "Hello, world" Spring Boot app that will be
used for this tutorial.

> Install ``devpack-for-spring`` and Java.

Author: https://github.com/canonical/devpack-for-spring/pull/84

> Install ``devpack-for-spring`` and Java.

Author: @erinecon this should be working by now

``spring-boot-framework`` profile, Rockcraft automates the creation of
``rockcraft.yaml`` and tailors the file for a Spring Boot app.
From the ``~/spring-boot-hello-world`` directory, initialize the rock:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:create-rockcraft-yaml]
    :end-before: [docs:create-rockcraft-yaml-end]
    :dedent: 2

The ``rockcraft.yaml`` file will automatically be created and set the name
based on your working directory.

Check out the contents of ``rockcraft.yaml``:

.. code-block:: bash

    cat rockcraft.yaml

The top of the file should look similar to the following snippet:

.. code-block:: yaml
    :caption: ~/spring-boot-hello-world/rockcraft.yaml

    name: spring-boot-hello-world
    # see https://documentation.ubuntu.com/rockcraft/en/latest/explanation/bases/
    # for more information about bases and using 'bare' bases for chiselled rocks
    base: bare # as an alternative, a ubuntu base can be used
    build-base: ubuntu@24.04 # build-base is required when the base is bare
    version: '0.1' # just for humans. Semantic versioning is recommended

>     summary: A summary of your Spring Boot application # 79 char long summary

Author: This should be fixed in https://github.com/canonical/rockcraft/pull/904

    description: |
        This is spring-boot-hello-world's description. You have a paragraph or two to tell the
        most important story about it. Keep it under 100 words though,
        we live in tweetspace and your description wants to look good in the
        container registries out there.
    # the platforms this rock should be built on and run on.
    # you can check your architecture with `dpkg --print-architecture`
    platforms:
        amd64:
        # arm64:
        # ppc64el:
        # s390x:

Verfiy that the ``name`` is ``spring-boot-hello-world``.

The ``platforms`` key must match the architecture of your host. Check
the architecture of your system:

.. code-block:: bash

    dpkg --print-architecture

Edit the ``platforms`` key in ``rockcraft.yaml`` if required.

> Now let's pack the rock:

Reviewer: This is as far as I got on my AMD64 Canonistack instance.

```
ubuntu@charm-dev:~/spring-boot-hello-world$ time rockcraft pack
*EXPERIMENTAL* extension 'spring-boot-framework' enabled
*EXPERIMENTAL* extension 'spring-boot-framework' enabled
Failed to cut requested chisel slices: libnss3_nss, libc6_libs, libgcc-s1_libs, libstdc++6_libs, zlib1g_libs
Detailed information: :: error: cannot talk to release repository: Get "https://codeload.github.com/canonical/chisel-releases/tar.gz/refs/heads/ubuntu-24.04": net/http: TLS handshake timeout
Failed to run rockcraft in instance
```

Reviewer: Could be, I was on the VPN when for this test (I was on a Canonistack instance)

Author: Retesting it with arm64, I managed to pack in 5m25s in a raspberry pi 5 :) so it had to be issues with the archive

    App                      Version  Status  Scale  Charm                    Channel  Rev  Address         Exposed  Message
    spring-boot-hello-world           active      1  spring-boot-hello-world             0  10.152.183.157  no

    Unit                        Workload  Agent  Address       Ports  Message
    spring-boot-hello-world/0*  active    idle   10.1.223.117

Let's expose the app using ingress. Deploy the
``nginx-ingress-integrator`` charm and integrate it with the Spring Boot app:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:deploy-nginx]
    :end-before: [docs:deploy-nginx-end]
    :dedent: 2

The hostname of the app needs to be defined so that it is accessible via
the ingress. We will also set the default route to be the root endpoint:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:config-nginx]
    :end-before: [docs:config-nginx-end]
    :dedent: 2

.. note::

    By default, the port for the Spring Boot app should be 8080. If you want to change
    the default port, it can be done with the configuration option ``app-port`` that

>     will be exposed as the ``SERVER_PORT`` to the Spring Boot app.

Reviewer: ```suggestion
    will be exposed as the environment variable ``SERVER_PORT`` to the Spring Boot app.
```
Sorry if this suggestion causes the lint test to fail :(

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:change-config]
    :end-before: [docs:change-config-end]
    :dedent: 2

After we wait for a moment for the app to be restarted, using
``curl http://spring-boot-hello-world  --resolve spring-boot-hello-world:80:127.0.0.1``
should now return the updated ``Hi!`` greeting.

Integrate with a database
-------------------------

Now let's keep track of how many visitors your app has received.
This will require integration with a database to keep the visitor count.
This will require a few changes:

- We will need to create a database migration that creates the ``visitors`` table.
- We will need to keep track how many times the root endpoint has been called
  in the database.
- We will need to add a new endpoint to retrieve the number of visitors from
  the database.

Let's start with the database migration to create the required tables.
We will use the auto DDL generation from JPA using the
property ``spring.jpa.generate-ddl``.

Go back out to the ``~/spring-boot-hello-world`` directory using ``cd ..``.
Add the following line to the ``src/main/resources/application.properties``
file:

> .. code-block:: Properties
> 
>     spring.jpa.generate-ddl=true

Reviewer: ```suggestion
.. code-block:: Properties
    :caption: ~/spring-boot-hello-world/src/main/resources/application.properties

    spring.jpa.generate-ddl=true
```

To connect the Spring Boot app to PostgreSQL, we will need
Spring Data JPA, the ``postgresl`` driver. We will also need the ``h2`` driver

> for the tests to pass. Add the following snippet in the ``pom.xml``

Reviewer: For my understanding: What tests?

under the ``dependencies`` tag:

> .. literalinclude:: code/spring-boot/pom.xml.visitors.dependencies.txt
>     :language: xml

Reviewer: ```suggestion
.. literalinclude:: code/spring-boot/pom.xml.visitors.dependencies.txt
    :caption: ~/spring-boot-hello-world/pom.xml
    :language: xml
```

Replace the code into
``src/main/java/com/example/demo/HelloController.java`` with the following:

.. literalinclude:: code/spring-boot/HelloController.java.visitors.txt
    :caption: ~/spring-boot-hello-world/src/main/java/com/example/demo/\
              HelloController.java
    :language: java

> Create the classes ``ApplicationConfig.java``, ``Visitor.java``,
> ``VisitorRepository.java``, ``VisitorService.java`` in the
> ``src/main/java/com/example/demo/ApplicationConfig.java`` directory
> with the following:

Reviewer: Personally I don't like having four code blocks back-to-back. I suggest updating this text to something like:
```suggestion
Now we'll create some new classes in the
``src/main/java/com/example/demo/`` directory.
Create the class ``ApplicationConfig`` in a new file
``ApplicationConfig.java`` with the following content:
```
And then similar sentences between the ``Visitor.java``, ``VisitorRepository.java``, and ``VisitorService.java`` code blocks.

.. literalinclude:: code/spring-boot/visitors_charmcraft.yaml
    :language: yaml

We can now pack and deploy the new version of the Spring Boot app:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:refresh-2nd-deployment]
    :end-before: [docs:refresh-2nd-deployment-end]
    :dedent: 2

Now let's deploy PostgreSQL and integrate it with the Spring Boot app:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:deploy-postgres]
    :end-before: [docs:deploy-postgres-end]
    :dedent: 2

Wait for ``juju status`` to show that the App is ``active`` again.
During this time, the Spring Boot app may enter a ``blocked`` state as it
waits to become integrated with the PostgreSQL database. Due to the
``optional: false`` key in the endpoint definition, the Spring Boot app will not
start until the database is ready.

Running
``curl http://spring-boot-hello-world --resolve spring-boot-hello-world:80:127.0.0.1``
should still return the ``Hi!`` greeting.

To check the local visitors, use
``curl http://spring-boot-hello-world/visitors --resolve spring-boot-hello-world:80:127.0.0.1``,

> which should return ``Number of visitors 1`` after the

Reviewer: Based on the ``HelloController.java`` file, shouldn't the output be `Number of visitors: 1` with a colon?

previous request to the root endpoint.
This should be incremented each time the root endpoint is requested. If we
repeat this process, the output should be as follows:

.. terminal::
    :input: curl http://spring-boot-hello-world --resolve spring-boot-hello-world:80:127.0.0.1

    Hi!
    :input: curl http://spring-boot-hello-world/visitors --resolve spring-boot-hello-world:80:127.0.0.1

>     Number of visitors 2

Reviewer: Same question as above; shouldn't the output be:
```suggestion
    Number of visitors: 2
```

Author: I have removed the `:` in the code instead, to be consistent with the go app.

Tear things down
----------------

We've reached the end of this tutorial. We went through the entire
development process, including:

- Creating a Spring Boot app
- Deploying the app locally
- Packaging the app using Rockcraft
- Building the app with Ops code using Charmcraft

> - Deplyoing the app using Juju

Reviewer: ```suggestion
- Deploying the app using Juju
```

- Exposing the app using an ingress
- Configuring the app
- Integrating the app with a database

If you'd like to quickly tear things down, start by exiting the Multipass VM:

.. code-block:: bash

    exit

And then you can proceed with its deletion:

.. code-block:: bash

    multipass delete charm-dev
    multipass purge

If you'd like to manually reset your working environment, you can run the

> following in the rock directory ``~/spring-boot-hello-world/charm`` for the tutorial:

Reviewer: ```suggestion
following in the directory ``~/spring-boot-hello-world/charm`` for the tutorial:
```

   Kubernetes cloud with Juju.
#. Repeat the process, mimicking a real development process.

.. important::

    Should you get stuck or notice issues, please get in touch on
    `Matrix <https://matrix.to/#/#12-factor-charms:ubuntu.com>`_ or
    `Discourse <https://discourse.charmhub.io/>`_

Set things up
-------------

.. include:: /reuse/tutorial/setup_edge.rst
.. |12FactorApp| replace:: Spring Boot

As the ``spring-boot-framework`` extensions for Rockcraft and Charmcraft are
still in development, we must enable experimental extensions for each:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:export-experimental-env-vars]
    :end-before: [docs:export-experimental-env-vars-end]
    :dedent: 2

Create the Spring Boot app
--------------------------

Start by creating the "Hello, world" Spring Boot app that will be
used for this tutorial.

Install ``devpack-for-spring`` and Java.

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:install-devpack-for-spring]
    :end-before: [docs:install-devpack-for-spring-end]
    :dedent: 2

Create the Demo Spring Boot application that will be used for

> this tutorial.

Author: I am not sure about that. I think a Spring Boot developer will understand all arguments (see [devpack-for-spring](https://github.com/canonical/devpack-for-spring/tree/main/devpack-for-spring) and [Spring Boot CLI](https://docs.spring.io/spring-boot/cli/using-the-cli.html))

> this tutorial.

Reviewer: Okay cool, thanks for the additional context. At the very least, I think we could add a `seealso` admonition and link the [Spring Boot CLI](https://docs.spring.io/spring-boot/cli/using-the-cli.html) page, but I'm also fine leaving this text as-is for now.

Deploy the Spring Boot app
--------------------------

A Juju model is needed to handle Kubernetes resources while deploying
the Spring Boot app. The Juju model holds the app along with any supporting
components. In this tutorial, our model will hold the Spring Boot app, ingress,
and a PostgreSQL database.

Let's create a new model:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:add-juju-model]
    :end-before: [docs:add-juju-model-end]
    :dedent: 2

Constrain the Juju model to your architecture:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:add-model-constraints]
    :end-before: [docs:add-model-constraints-end]
    :dedent: 2

Now let's use the OCI image we previously uploaded to deploy the Spring Boot
app. Deploy using Juju by specifying the OCI image name with the
``--resource`` option:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:deploy-app]
    :end-before: [docs:deploy-app-end]
    :dedent: 2

It will take a few minutes to deploy the Spring Boot app. You can monitor its
progress with:

.. code-block:: bash

>     juju status --watch 2s

Reviewer: Okay, I'll give it another try without adding `paas-charm` to the `requirements.txt` file and see what happens

>     juju status --watch 2s

Reviewer: Yes, that was the trick! I successfully deployed the Spring Boot app :D thanks!

    version: '0.2' # just for humans. Semantic versioning is recommended
    summary: A summary of your Spring Boot app # 79 char long summary
    description: |
        This is spring-boot-hello-world's description. You have a paragraph or two to tell the
        most important story about it. Keep it under 100 words though,
        we live in tweetspace and your description wants to look good in the
        container registries out there.
    # the platforms this rock should be built on and run on.
    # you can check your architecture with `dpkg --print-architecture`
    platforms:
        amd64:
        # arm64:
        # ppc64el:
        # s390x:

Let's pack and upload the rock:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:skopeo-update]
    :end-before: [docs:skopeo-update-end]
    :dedent: 2

Change back into the charm directory using ``cd charm``.

The ``spring-boot-framework`` Charmcraft extension supports adding configurations
to ``charmcraft.yaml``, which will be passed as environment variables to
the Spring Boot app. Add the following to the end of the
``charmcraft.yaml`` file:

> .. literalinclude:: code/spring-boot/greeting_charmcraft.yaml
>     :language: yaml

Reviewer: ```suggestion
.. literalinclude:: code/spring-boot/greeting_charmcraft.yaml
    :caption: ~/spring-boot-hello-world/charm/charmcraft.yaml
    :language: yaml
```

    version: '0.3' # just for humans. Semantic versioning is recommended
    summary: A summary of your Spring Boot app # 79 char long summary
    description: |
        This is spring-boot-hello-world's description. You have a paragraph or two to tell the
        most important story about it. Keep it under 100 words though,
        we live in tweetspace and your description wants to look good in the
        container registries out there.
    # the platforms this rock should be built on and run on.
    # you can check your architecture with `dpkg --print-architecture`
    platforms:
        amd64:
        # arm64:
        # ppc64el:
        # s390x:

Let's pack and upload the rock:

.. literalinclude:: code/spring-boot/task.yaml
    :language: bash
    :start-after: [docs:skopeo-2nd-update]
    :end-before: [docs:skopeo-2nd-update-end]
    :dedent: 2

Change back into the charm directory using ``cd charm``.

The Spring Boot app now requires a database which needs to be declared in the
``charmcraft.yaml`` file. Open ``charmcraft.yaml`` in a text editor and
add the following section to the end of the file:

> .. literalinclude:: code/spring-boot/visitors_charmcraft.yaml
>     :language: yaml

Reviewer: ```suggestion
.. literalinclude:: code/spring-boot/visitors_charmcraft.yaml
    :caption: ~/spring-boot-hello-world/charm/charmcraft.yaml
    :language: yaml
```
